### PR TITLE
Add OpenCL 3.0 deviceInfo capabilities

### LINKED ISSUE #796: OpenCL 3.0 - computeinfo test - new capabilities
I am going to extend computeinfo tests to new deviceinfo capabilities

CL_DEVICE_ATOMIC_MEMORY_CAPABILITIES,
CL_DEVICE_ATOMIC_FENCE_CAPABILITIES,
CL_DEVICE_NON_UNIFORM_WORK_GROUP_SUPPORT
CL_DEVICE_PREFERRED_WORK_GROUP_SIZE_MULTIPLE
CL_DEVICE_WORK_GROUP_COLLECTIVE_FUNCTIONS_SUPPORT,
CL_DEVICE_GENERIC_ADDRESS_SPACE_SUPPORT
CL_DEVICE_OPENCL_C_FEATURES
CL_DEVICE_DEVICE_ENQUEUE_SUPPORT
CL_DEVICE_PIPE_SUPPORT

### PR DESCRIPTION
**Pull request by gwawiork**
Fixes #796 
Full list of supported capabilities by this commit:
CL_DEVICE_ATOMIC_MEMORY_CAPABILITIES,
CL_DEVICE_ATOMIC_FENCE_CAPABILITIES,
CL_DEVICE_NON_UNIFORM_WORK_GROUP_SUPPORT,
CL_DEVICE_PREFERRED_WORK_GROUP_SIZE_MULTIPLE,
CL_DEVICE_WORK_GROUP_COLLECTIVE_FUNCTIONS_SUPPORT,
CL_DEVICE_GENERIC_ADDRESS_SPACE_SUPPORT,
CL_DEVICE_OPENCL_C_FEATURES,
CL_DEVICE_DEVICE_ENQUEUE_SUPPORT,
CL_DEVICE_PIPE_SUPPORT,
CL_DEVICE_NUMERIC_VERSION,
CL_DEVICE_EXTENSIONS_WITH_VERSION,
CL_DEVICE_OPENCL_C_ALL_VERSIONS,
CL_DEVICE_ILS_WITH_VERSION,
CL_DEVICE_BUILT_IN_KERNELS_WITH_VERSION

### CODE REVIEW

**Comment by gwawiork** [comment]
Example output:

CL_DEVICE_ATOMIC_MEMORY_CAPABILITIES == CL_DEVICE_ATOMIC_ORDER_RELAXED|CL_DEVICE_ATOMIC_ORDER_ACQ_REL|CL_DEVICE_ATOMIC_ORDER_SEQ_CST	CL_DEVICE_ATOMIC_FENCE_CAPABILITIES == CL_DEVICE_ATOMIC_ORDER_RELAXED|CL_DEVICE_ATOMIC_ORDER_ACQ_REL|CL_DEVICE_ATOMIC_ORDER_SEQ_CST	CL_DEVICE_NON_UNIFORM_WORK_GROUP_SUPPORT == 1
	CL_DEVICE_PREFERRED_WORK_GROUP_SIZE_MULTIPLE == 64
	CL_DEVICE_WORK_GROUP_COLLECTIVE_FUNCTIONS_SUPPORT == 1
	CL_DEVICE_GENERIC_ADDRESS_SPACE_SUPPORT == 1
	CL_DEVICE_OPENCL_C_FEATURES supported name and version:
		"__opencl_c_atomic_order_acq_rel" 3.0.0
		"__opencl_c_3d_image_writes" 3.0.0
CL_DEVICE_DEVICE_ENQUEUE_SUPPORT == 1
	CL_DEVICE_PIPE_SUPPORT == 1
	CL_DEVICE_NUMERIC_VERSION == 3.0.0
	CL_DEVICE_EXTENSIONS_WITH_VERSION supported name and version:
		"cl_khr_byte_addressable_store" 1.0.0
		"cl_khr_fp16" 1.0.0
CL_DEVICE_OPENCL_C_ALL_VERSIONS supported name and version:
		"OpenCL C" 1.0.0
		"OpenCL C" 1.1.0
		"OpenCL C" 1.2.0
CL_DEVICE_ILS_WITH_VERSION supported name and version:
		"SPIR-V_1.2 " 1.2.0
CL_DEVICE_BUILT_IN_KERNELS_WITH_VERSION supported name and version:
		""

**Comment by gwawiork** [comment]
Please, do not merge yet. Need changes around CL_DEVICE_DEVICE_ENQUEUE_SUPPORT after documentation update https://github.com/KhronosGroup/OpenCL-Docs/pull/321

**Comment by gwawiork** [comment]
Added support against https://github.com/KhronosGroup/OpenCL-Headers/pull/96 and KhronosGroup/OpenCL-Docs#321

**Comment by gwawiork** [comment]
I am working on adding CL_PLATFORM_EXTENSIONS and CL_PLATFORM_EXTENSIONS_WITH_VERSION in this PR

**Comment by gwawiork** [comment]
CL_PLATFORM_EXTENSIONS and CL_PLATFORM_EXTENSIONS_WITH_VERSION have been added.
I think all OpenCL 3.0 capabilities are covered. Please review

**Comment by gwawiork** [comment]
Building has failed bacause https://github.com/KhronosGroup/OpenCL-Headers/pull/96 not merged yet.

File: test_conformance/computeinfo/main.cpp [hunk 1]
```diff
@@ -370,6 +396,62 @@ int getImageInfo(cl_device_id device, const version_t& version)
...
+            }
+            size_err = config_size_set != config_size_ret;
+            break;
+        default:
+            log_error("Unknown config type: %d\n", info->config_type);
+            break;
+    }
+    if (err || size_err)
+        log_error("\tFailed clGetPlatformInfo for %s.\n", info->opcode_name);
+    if (err) print_error(err, "\t\clGetPlatformInfo failed.");
```
Comment (bashbaug) [inline test_conformance/computeinfo/main.cpp:451]: I was able to re-trigger the Travis build and this line is generating an error:

```
/Users/travis/build/KhronosGroup/OpenCL-CTS/test_conformance/computeinfo/main.cpp:451:34: error: unknown escape sequence '\c' [-Werror,-Wunknown-escape-sequence]
    if (err) print_error(err, "\t\clGetPlatformInfo failed.");
```

Fix is straightforward:

```suggestion
    if (err) print_error(err, "\t\tclGetPlatformInfo failed.");
```

File: test_conformance/computeinfo/main.cpp [hunk 2]
```diff
@@ -480,8 +599,9 @@ int getConfigInfo(cl_device_id device, config_info* info)
     return err || size_err;
 }
 
-void dumpConfigInfo(cl_device_id device, config_info* info)
+void dumpConfigInfo(config_info* info)
 {
+    int number_of_version_items = 0;
```
Comment (EwanC) [inline test_conformance/computeinfo/main.cpp:604]: Looks like this is only used in one case of the switch, could declare it there on line 899?

**Comment by bashbaug** [comment]
As there are two approvals for this PR, I am merging these changes.

**Pull request merged**